utils/exceptions.hpp

```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include <QException>
+#include <QMessageBox>
+
+namespace jbrush
+{
+// Inherits from QException and allows the client to specify a unique string message that
+// is displayed in Qt's debugging console when raised
+class GeneralException : public QException
+{
+private:
+    QString message;
+public:
+    GeneralException(QString m) { message = m; }
+    void raise() const override { throw *this; }
+    GeneralException* clone() const override { return new GeneralException(*this); }
+
+    // Return the error message assigned to this exception
+    QString errorMessage () const { return message; }
+
+    // Causes a message box to appear describing the error that occurred
+    // Good for informing the user of errors they need to be aware of
+    void errorWindow () const
+    {
+       QMessageBox::critical(nullptr, "Error", message);
+    }
+};
+
+// Throw this exception if a file was not opened properly
+class BadFile : public GeneralException
+{
+public:
+    BadFile(QString m) : GeneralException(m) {}
+};
+// Throw this exception if the file attempted to open did not have the correct format
+class BadFileFormat : public GeneralException
```

Comment:
This class should be specific to bad file formatting. It should require the line number, what it expected, and what it obtained. Then we can properly give an informational error message.
```
Line 51: Expected a number. Obtained a ",".
```

Reply:
Do you think the line number should refer to the line number in the file or in the code?

Reply:
actually nvm, it looks like there is no convenient way to get the current line number of a file just from the class alone